### ISSUE TITLE
Problems with basename

### ISSUE THREAD

**Issue by olmobrutall** [ReactTraining/react-router#5000]
Hi guys, 

Congrats for the v4 release!

I've problems using `basename`. 

The basename is prepended or removed depending on the use case.

```
basename: "MyApp"
location: "/MyModule" 
finalUrl: /MyApp/MyModule
```

But there is no way to know if a path is app-relative or an absolute url, so calling `createHref` twice duplicates the basename: 

```JS
var url = history.createHref("/MyModule"); //Returns /MyApp/MyModule

//later
if(e.ctrl || e.button == 1)
   window.open(url); //works
else
   history.push(url); //duplicated here /MyApp/MyApp/MyModule
```

Is there any simple pattern to solve this? 

In .Net for example an app-relative url will start with tilde `~/MyModule` and gets converted to `/MyApp/MyModule`. This solves the problem of duplicate base names and adds clarity. 

I'm not suggesting to adopt the .Net solution here, but will be awesome if there will be two new extension points in history object. 

* `toAbsoluteUrl(appRelativeUrl)`, used by `push`, `replace` and `createHref`. 
By default `return basename +  appRelativeUrl`  
* `toAppRelative(absoluteUrl)`, used by `matchPath` so it affects o all the routes. 
By default `return absoluteUrl.removePrefix(basename)`

With this extension points I could have one consistent way for all app-relative urls, independently if they are inside the SPA or in the server-side. 

Thanks for your time

**Comment by timdorr** [comment]
I think there's a fundamental misunderstanding about the basename here. It comes from the `<base href="/foo">` HTML element, which applies a basename to all links on the page. In that case, the rest of the HTML doesn't "know" about that base URL; it's just automatically prepended to any href. 

Similarly, the basename in Router is something the application doesn't know about. It just `<Link>`'s to various routes and is unaware of the basename being attached. 

One thing to point out here is you should very rarely be using `createHref`, and never be analyzing/parsing the URL it returns. It's assumed to be a serialized/stringified form of the location object provided to it. It accepts a string, but that is a convenience feature. Try to think of all the history methods in terms of location objects (with `pathname`, `search`, etc. properties).

And back to your specific problem, you should instead be swapping out how you form those links. You don't care how the app's URLs are based, just that they link to valid routes. If you need to link outside of the app, just use a plain `<a>` element.

**Comment by olmobrutall** [comment]
Both `<base href/>`, `basename` and `~/` solution try to fix the problem that HTML didn't provide a concept of `app` different to `domain`.  

So instead of writing `<Link to="~/myModule">` and `<Route path="~/myModule">` 
I should write `<Link to={fixUrl("~/myModule")}>` and `<Route path={fixUrl("~/myModule")}>`  in every single Link call to push or Route?  

That will work but it's quite annoying. For now I'm monkey-patching history, Route, and Switch. It's not beautiful but at least is isolated. Will be awesome to have some extension points to make it more elegant.

**Comment by eloytoro** [comment]
This is a regression @timdorr react-router v3 makes it so you can have dynamic basenames without modifying the html by passing it as a parameter to the `createBrowserHistory` function

```jsx
const history = createBrowserHistory({ basename: '/foo' });

<Link to="bar" /> // this would lead to /foo/bar in v3, but not in v4
```

**Comment by eloytoro** [comment]
^ was fixed in #5110
